05_debug/procfs/task_05_procfs.c

```diff
@@ -0,0 +1,292 @@
+		remove_proc_entry(TEXT_BUF_FILENAME, mod_dir);
+		text_buf_file = NULL;
+	}
+
+	if (mod_dir) {
+		remove_proc_entry(PROC_DIRECTORY, NULL);
+		mod_dir = NULL;
+	}
+}
+
+///////////////////////////////////////////////////////////////////////////////
+
+static ssize_t text_buf_read(struct file *file_p, char __user *buffer, size_t length, loff_t *offset)
+{
+	return simple_read_from_buffer(buffer, length, offset, text_buf, strlen(text_buf));
+}
+
+///////////////////////////////////////////////////////////////////////////////
+
+static int text_buf_write(struct file *file_p, const char __user *buffer, size_t length, loff_t *offset)
+{
+	size_t left;
+
+	text_buf = kzalloc(sizeof(*text_buf) * 100, GFP_KERNEL);
+	left = copy_from_user(text_buf, buffer, 100);
+
+	if (left)
+		pr_err("failed to write %u from %u chars\n", left, length);
+	else
+		pr_info("Word '%s' has converted to UPPER case successfuly\n", text_buf);
+
+	text_buf = change_case(text_buf, LOW_TO_UP);
+
+	total_characters_processed += characters_converted;
+	total_calls++;
+
+	return length;
+}
+
+///////////////////////////////////////////////////////////////////////////////
+
+static int total_calls_read(struct file *file_p, char __user *buffer, size_t length, loff_t *offset)
+{
+	char tmp[16];
+
+	sprintf(tmp, "%d\n", total_calls);
+
+	return simple_read_from_buffer(buffer, length, offset, tmp, strlen(tmp));
+}
+
+///////////////////////////////////////////////////////////////////////////////
+
+static int total_characters_processed_read(struct file *file_p, char __user *buffer, size_t length, loff_t *offset)
+{
+	char tmp[16];
+
+	sprintf(tmp, "%d\n", total_characters_processed);
+
+	return simple_read_from_buffer(buffer, length, offset, tmp, strlen(tmp));
+}
+
+///////////////////////////////////////////////////////////////////////////////
+
+static int characters_converted_read(struct file *file_p, char __user *buffer, size_t length, loff_t *offset)
+{
+	char tmp[16];
+
+	sprintf(tmp, "%d\n", characters_converted);
+
+	return simple_read_from_buffer(buffer, length, offset, tmp, strlen(tmp));
+}
+
+///////////////////////////////////////////////////////////////////////////////
+//###########################INIT CALLBACK#####################################
+///////////////////////////////////////////////////////////////////////////////
+
+static int __init mod_init(void)
+{
+	int err;
+
+	text_buf = kstrdup("Buffer is empty:(\n", GFP_KERNEL);
+
+	err = _create_proc();
+	if (err)
+		goto error;
+
+	pr_info("'task_05' module initialized\n");
+	return 0;
+
+error:
+	pr_err("'task_05' module failed to load\n");
+	_cleanup_proc();
+	return err;
+}
+
+///////////////////////////////////////////////////////////////////////////////
+//###########################EXIT CALLBACK#####################################
+///////////////////////////////////////////////////////////////////////////////
+
+static void __exit mod_exit(void)
+{
+	_cleanup_proc();
+
+	if (text_buf != NULL)
```

Comment:
It is possible to call kfree with NULL argument. If argument is NULL, no operation is performed.